File: problems/443/443.string-compression.py
#
# @lc app=leetcode id=443 lang=python3
#
# [443] String Compression
#


# @lc code=start
class Solution:
    def compress(self, chars: List[str]) -> int:
        # modify in place
        # return len of new arr

        n = len(chars)
        curr_i = 0
        curr_char = chars[0]
        curr_cnt = 1

        for i in range(1, n):
            if chars[i] != curr_char:
                chars[curr_i] = curr_char
                curr_i += 1
                if curr_cnt > 1:
                    curr_cnt = str(curr_cnt)
                    for digit in curr_cnt:
                        chars[curr_i] = digit
                        curr_i += 1
                curr_char = chars[i]
                curr_cnt = 1
            else:
                curr_cnt += 1

        chars[curr_i] = curr_char
        curr_i += 1
        if curr_cnt > 1:
            curr_cnt = str(curr_cnt)
            for digit in curr_cnt:
                chars[curr_i] = digit
                curr_i += 1

        return curr_i


# @lc code=end

File: problems/443/443.string-compression.java
/*
 * @lc app=leetcode id=443 lang=java
 *
 * [443] String Compression
 */

// @lc code=start
class Solution {
    public int compress(char[] chars) {
        // intuition:
        // keep track of count and update point k
        // k never over takes i
        // O(N) time, O(1) space

        int k = 0;
        int i = 0;

        while (i < chars.length) {
            // iterate till we find char not equal to prev or till end of chars
            int j = i;
            while (j < chars.length && chars[j] == chars[i]) {
                j++;
            }

            // now compress chars
            chars[k++] = chars[i];
            if (j - i > 1) {
                String cnt = j - i + "";
                for (char c : cnt.toCharArray()) {
                    chars[k++] = c;
                }
            }

            i = j;
        }

        return k;
    }
}
// @lc code=end
